src/cure.sol

```diff
@@ -0,0 +1,115 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+/// cure.sol -- Debt Rectifier contract
+
+// Copyright (C) 2022 Dai Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.6.12;
+
+interface SourceLike {
+    function cure() external view returns (uint256);
+}
+
+contract Cure {
+    mapping (address => uint256) public wards;
+    uint256 public live;
+    address[] public sources;
+    uint256 public total;
+    mapping (address => uint256) public pos; // position in sources + 1, 0 means a source does not exist
+    mapping (address => uint256) public amt;
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event Cage();
+
+    modifier auth {
+        require(wards[msg.sender] == 1, "Cure/not-authorized");
+        _;
+    }
+
+    modifier isLive {
+        require(live == 1, "Cure/not-live");
+        _;
+    }
+
+    // --- Internal ---
+    function _add(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x + y) >= x, "Cure/add-overflow");
+    }
+
+    function _sub(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x - y) <= x, "Cure/sub-underflow");
+    }
+
+    constructor() public {
+        live = 1;
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    function count() external view returns (uint256 count_) {
+        count_ = sources.length;
+    }
+
+    function list() external view returns (address[] memory) {
+        return sources;
+    }
+
+    function rely(address usr) external auth isLive {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+
+    function deny(address usr) external auth isLive {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    function addSource(address src) external auth isLive {
+        require(pos[src] == 0, "Cure/already-existing-source");
+        sources.push(src);
+        pos[src] = sources.length;
+        uint256 amt_ = amt[src] = SourceLike(src).cure();
```

Comment:
Either way is probably ok, but I would remove the cure sampling here, thus separating the logic of adding  a source and sampling it. Just seems to me a bit cleaner.
If we want to sample (/reset) as part of the spell we can just add that call right after.

Reply:
Yes you have a good point, however I'd prefer to leave this one open for now to see others' opinions before applying the change.

Reply:
Yeah I think I'm leaning more towards leaving it out to separate the logic. `reset()` will only need to be called after shutdown really.

Reply:
I don't have terribly strong feelings here, it's almost just a matter of aesthetics. Might personally lean towards a call to `reset` in `addSource` to avoid duplicating logic but quickly confirming that the added source supports the correct API. But really I think any option will work.